tests/modules/blockedclient.c

```diff
@@ -250,32 +271,30 @@ int do_rm_call_async(RedisModuleCtx *ctx, RedisModuleString **argv, int argc){
         return RedisModule_WrongArity(ctx);
     }
 
-    RedisModuleBlockedClient *bc = NULL;
-    RedisModuleOnUnblocked on_unblock = NULL;
-    int flags = RedisModule_GetContextFlags(ctx);
-    if (!(flags & REDISMODULE_CTX_FLAGS_DENY_BLOCKING)) {
-        bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
-        on_unblock = rm_call_async_on_unblocked;
+    sds format = sdsempty();
+
+    if (!(RedisModule_GetContextFlags(ctx) & REDISMODULE_CTX_FLAGS_DENY_BLOCKING)) {
+        /* We are allowed to block the client so we can allow RM_Call to also block us */
+        format = sdscat(format, "K");
     }
 
-    int script_mode = 0;
     const char* invoked_cmd = RedisModule_StringPtrLen(argv[0], NULL);
     if (strcasecmp(invoked_cmd, "do_rm_call_async_script_mode") == 0) {
-        script_mode = 1;
+        format = sdscat(format, "S");
     }
 
+    format = sdscat(format, "Ev");
+
     const char* cmd = RedisModule_StringPtrLen(argv[1], NULL);
 
-    RedisModuleCallReply* rep = RedisModule_Call(ctx, cmd, script_mode ? "SEKv" : "EKv", argv + 2, argc - 2);
+    RedisModuleCallReply* rep = RedisModule_Call(ctx, cmd, format, argv + 2, argc - 2);
+    sdsfree(format);
+
     if(RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_PROMISE) {
-        RedisModuleCtx *bctx = bc? RedisModule_GetThreadSafeContext(bc) : ctx;
-        rm_call_async_send_reply(bctx, rep);
-        if (bc) {
-            RedisModule_FreeThreadSafeContext(bctx);
-            RedisModule_UnblockClient(bc, NULL);
-        }
+        rm_call_async_send_reply(ctx, rep);
     } else {
-        RedisModule_CallReplyPromiseSetUnblockHandler(rep, on_unblock, bc);
+        RedisModuleBlockedClient *bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);;
```

Comment:
you can do the same thing (of not creating the blocked client in advance) in the other function as well.